README.md
# node-lightify-rest
[![](https://img.shields.io/npm/dt/node-lightify-rest.svg?maxAge=2592000)](https://www.npmjs.com/package/node-lightify-rest) [![](https://img.shields.io/npm/v/node-lightify-rest.svg?maxAge=2592000)](https://www.npmjs.com/package/node-lightify-rest) [![](https://www.versioneye.com/user/projects/579f48ef72d75c0039f7a287/badge.svg)](https://www.versioneye.com/user/projects/579f48ef72d75c0039f7a287) [![](https://img.shields.io/travis/maxcanna/node-lightify-rest.svg?maxAge=2592000)](https://travis-ci.org/maxcanna/node-lightify-rest/) [![](https://img.shields.io/codeclimate/github/maxcanna/node-lightify-rest.svg?maxAge=2592000)](https://codeclimate.com/github/maxcanna/node-lightify-rest) [![](https://img.shields.io/npm/l/node-lightify-rest.svg?maxAge=2592000)](https://github.com/maxcanna/node-lightify-rest/blob/master/LICENSE)

Node.js library to use [Osram Lightify REST API](https://eu.lightify-api.org/lightify/)

## Install

``$ npm i node-lightify-rest``

## Usage

```javascript
var Lightify = require('node-lightify-rest');

var lightify = new Lightify({
    'username': 'user@example.com',
    'password': 'thisisaverylongandsecurepassword',
    'serial': 'OSR0000000A',
    'region': Lightify.Regions.eu, //For Europe or Lightify.Regions.us for US, CA, AUS
});

lightify
    .getDevices()
    .then(console.log)
    .catch(console.error);
```

## Documentation
[Osram Lightify REST API official documentation](https://eu.lightify-api.org/lightify/)

### Parameters
| Name | Data type | Description | Example
|------|-----------|-------------|--------|
| apiVersion | string | API version | 1.0.0 |
| color | string | Hex Color | FF0000 |
| ctemp | number | Color temperature | 1000 to 8000 |
| hue | number | Hue | 1.000 to 360.000 |
| idx | number | Index of the device | 1 |
| level | number | Dimming level | 0.000 to 1.000 |
| onoff | number | On/Off state | 0 or 1 |
| recallSceneId | string | Recall Scene Id | 1, 2-1 |
| saturation | number | Saturation | 0.000 to 1.000 |
| time | number | Transition time in 1/10th of a second | 100 |

## API
All API methods are available using corresponding method of `Lightify` class.

[Examples](https://github.com/maxcanna/node-lightify-rest/tree/master/examples) may be found in `examples` directory.

## Contributing
Contributions are very welcome!

Note that submitting a PR you agree to license your contribution to this project under the [GPL License](https://github.com/maxcanna/node-lightify-rest/blob/master/LICENSE).

Code style are enforced using automated checks.

## Troubleshooting
Make sure that the firmware version of your gateway is `WLAN 1.1.2.101` or greater.

## License

Published under the [GPL License](https://github.com/maxcanna/node-lightify-rest/blob/master/LICENSE).


CHANGELOG.md
## Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

### v1.0.2 2016-08-28
- [#8](https://github.com/maxcanna/node-lightify-rest/pull/8) Edit code climate checks (@maxcanna)
- [#7](https://github.com/maxcanna/node-lightify-rest/pull/7) Use eslint (@maxcanna)
- [#6](https://github.com/maxcanna/node-lightify-rest/pull/6) Reduce code duplication (@maxcanna)
- [#5](https://github.com/maxcanna/node-lightify-rest/pull/5) Enable strict mode (@maxcanna)

### v1.0.0 2016-08-07
- [#3](https://github.com/maxcanna/node-lightify-rest/pull/3) Add Lightify class (@maxcanna)
- [#1](https://github.com/maxcanna/node-lightify-rest/pull/1) Add request handler (@maxcanna)


release.sh
#!/bin/sh

if [ -z "$(which github-changes)" ]; then
  # specify version because github-changes "is under heavy development. Things
  # may break between releases" until 0.1.0
  echo "First, do: [sudo] npm install -g github-changes@0.0.14"
  exit 1
fi

if [ -d .git/refs/remotes/upstream ]; then
  remote=upstream
else
  remote=origin
fi

# Increment v2.x.y -> v2.x+1.0
npm version minor || exit 1

# Generate changelog from pull requests
github-changes -o maxcanna \
  -r node-lightify-rest \
  --verbose \
  --file CHANGELOG.md \
  --only-pulls --use-commit-body \
  --date-format 'YYYY-MM-DD' \
  --title 'Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).' \
  || exit 1

# Since the tag for the new version hasn't been pushed yet, any changes in it
# will be marked as "upcoming"
version="$(grep '"version"' package.json | cut -d'"' -f4)"
sed -i -e "s/^### upcoming/### v$version -/" CHANGELOG.md

# This may fail if no changelog updates
git add CHANGELOG.md; git commit -m 'Update changelog'

# Publish the new version to npm
npm publish || exit 1

# Increment v2.x.0 -> v2.x.1
# For rationale, see:
# https://github.com/request/oauth-sign/issues/10#issuecomment-58917018
npm version patch || exit 1

# Push back to the main repo
git push $remote master --tags || exit 1


lib/request.js
/**
 * Created by massimilianocannarozzo on 01/08/16.
 */

"use strict";
const request = require('request-promise')
    , errors = require('request-promise/errors')
    , BPromise = require('bluebird');

class Request {
    constructor(options) {
        options = Object.assign({region: Request.Regions.eu}, options);

        if (Request.Regions[options.region] === undefined) {
            throw new Error('Invalid region');
        }

        if (options.username === undefined || options.password === undefined || options.serial === undefined) {
            throw new Error('Missing options');
        }

        this.region = options.region;
        this.username = options.username;
        this.password = options.password;
        this.serial = options.serial;
        this.token = '';
        this.request = request.defaults({
            baseUrl: `https://${this.region}.lightify-api.org/lightify/services/`,
            json: true
        });
    }

    _authenticate() {
        return new BPromise((resolve, reject) => {
            this.request.post({
                    url: 'session',
                    body: {
                        'username': this.username,
                        'password': this.password,
                        'serialNumber': this.serial,
                    },
                })
                .then((function (data) {
                    this.token = data['securityToken'];
                    resolve();
                }).bind(this))
                .catch(e => {
                    reject(e);
                });
        });
    }

    _requestData(options) {
        return this.request({
            method: options.method,
            qs: options.qs,
            url: options.url,
            headers: {
                'authorization': this.token,
                'Api-Version': '1.0.0',
            },
        });
    }

    doRequest(options) {
        return new BPromise((resolve, reject) => {
            this._requestData(options)
                .catch(errors.StatusCodeError, (function (e) {
                    if (e.response.body.errorCode === 5003) {
                        return this._authenticate();
                    }
                    reject(e.response.body.errorMessage);
                }).bind(this))
                .then((function (data) {
                    return data ? data : this._requestData(options);
                }).bind(this))
                .then(resolve)
                .catch(errors.StatusCodeError, e => {
                    reject(e.response.body.errorMessage);
                })
                .catch(errors.RequestError, reject);
        });
    }
}

Request.Regions = {
    eu: 'eu',
    us: 'us',
};

module.exports = Request;


examples/italyFlag.js
/**
 * Created by massimilianocannarozzo on 06/08/16.
 */

var Lightify = require('./../index');

var lightify = new Lightify({
    'username': '',
    'password': '',
    'serial': '',
    'region': Lightify.Regions.eu,
});

var green = {};
var white = {};
var red = {};

red[Lightify.Parameters.idx] = 3;
red[Lightify.Parameters.color] = 'CE2B37';
red[Lightify.Parameters.time] = 20;

white[Lightify.Parameters.idx] = 2;
white[Lightify.Parameters.color] = 'F1F2F1';
white[Lightify.Parameters.time] = 20;

green[Lightify.Parameters.idx] = 1;
green[Lightify.Parameters.color] = '009246';
green[Lightify.Parameters.time] = 20;

lightify
    .getDevices()
    .then(function (devices) {
        if (devices.count < 3) {
            throw new Error('This demo requires at least 3 devices');
        }
    })
    .then(function () {
        return lightify.turnOn();
    })
    .then(function () {
        return lightify.setDeviceParameters(green);
    })
    .then(function () {
        return lightify.setDeviceParameters(white);
    })
    .then(function () {
        return lightify.setDeviceParameters(red);
    })
    .catch(console.error);


index.js
/**
 * Created by massimilianocannarozzo on 04/08/16.
 */

"use strict";
const Request = require('./lib/request')
    , BPromise = require('bluebird');

const checkParameters = function (parameters) {
    Object.keys(parameters).forEach(parameter => {
        if (Lightify.Parameters[parameter] === undefined) {
            throw new Error('Invalid parameter ' + parameter);
        }
    });
};

class Lightify {
    constructor(options) {
        if (options.username === undefined || options.password === undefined || options.serial === undefined) {
            throw new Error('Missing options');
        }

        this.request = new Request({
            'username': options.username,
            'password': options.password,
            'serial': options.serial,
            'region': options.region,
        });
    }

    turnOn(idx) {
        return this.setDeviceParameters({idx: idx, onoff: 1});
    }

    turnOff(idx) {
        return this.setDeviceParameters({idx: idx, onoff: 0});
    }

    setDeviceParameters(parameters) {
        try {
            checkParameters(parameters);
        } catch (e) {
            return BPromise.reject(e.message);
        }
        return this.request.doRequest({
            method: 'GET',
            url: parameters.idx ? 'device/set' : 'device/all/set',
            qs: parameters,
        });
    }

    turnOnGroup(idx) {
        return this.setGroupParameters({idx: idx, onoff: 1});
    }

    turnOffGroup(idx) {
        return this.setGroupParameters({idx: idx, onoff: 0});
    }

    setGroupParameters(parameters) {
        try {
            checkParameters(parameters);
        } catch (e) {
            return BPromise.reject(e.message);
        }
        if (parameters.idx === undefined) {
            return BPromise.reject('Missing ' + Lightify.Parameters.idx);
        }

        return this.request.doRequest({
            method: 'GET',
            url: 'group/set',
            qs: parameters,
        });
    }

    getDevices() {
        return this.request.doRequest({
            method: 'GET',
            url: 'devices',
        });
    }

    getDevice(idx) {
        if (idx === undefined) {
            return BPromise.reject('Missing ' + Lightify.Parameters.idx);
        }

        return this.request.doRequest({
            method: 'GET',
            url: `devices/${idx}`,
        });
    }

    getGateway() {
        return this.request.doRequest({
            method: 'GET',
            url: 'gateway',
        });
    }

    getGroups() {
        return this.request.doRequest({
            method: 'GET',
            url: 'groups',
        });
    }

    getApiVersion() {
        return this.request.doRequest({
            method: 'GET',
            url: 'version',
        });
    }
}

Lightify.Regions = Request.Regions;

Lightify.Parameters = {
    idx: 'idx',
    color: 'color',
    ctemp: 'ctemp',
    hue: 'hue',
    level: 'level',
    onoff: 'onoff',
    saturation: 'saturation',
    time: 'time',
};

module.exports = Lightify;
